modules/gateway/nodes.go

```diff
@@ -102,68 +102,126 @@ func (g *Gateway) requestNodes(conn modules.PeerConn) error {
-				continue
-			}
-		}
-
-		// find an untested node to check
-		g.mu.RLock()
-		node, err := g.randomNode()
-		g.mu.RUnlock()
-		if err != nil {
+		if numNodes < pruneNodeListLen {
+			// There are not enough nodes in the gateway - pruning more is
+			// probably a bad idea, and may affect the user's ability to
+			// connect to the network in the future.
 			continue
 		}
 
-		// try to connect
-		conn, err := net.DialTimeout("tcp", string(node), dialTimeout)
+		// Try connecting to the random node. If the node is not reachable,
+		// remove them from the node list. Make sure that the dial is stopped
+		// early if the gateway is shutting down.
+		cancelDialChan := make(chan struct{})
+		dialDoneChan := make(chan struct{})
+		defer close(dialDoneChan)
+		dialer := &net.Dialer{
+			Timeout: dialTimeout,
+			Cancel:  cancelDialChan,
+		}
+		go func() {
+			select {
+			case <-cancelDialChan:
+			case <-g.threads.StopChan():
+				close(cancelDialChan)
+			}
+		}()
+		conn, err := dialer.Dial("tcp", string(node))
 		if err != nil {
 			g.mu.Lock()
 			g.removeNode(node)
 			g.save()
 			g.mu.Unlock()
 			g.log.Debugf("INFO: removing node %q because dialing it failed: %v", node, err)
-			continue
 		}
-		// if connection succeeds, supply an unacceptable version to ensure
-		// they won't try to add us as a peer
+		// If connection succeeds, supply an unacceptable version so that we
+		// will not be added as a peer.
+		//
+		// NOTE: this is a somewhat clunky way of specifying that you didn't
+		// actually want a connection.
 		encoding.WriteObject(conn, "0.0.0")
 		conn.Close()
-		// sleep for an extra 10 minutes after success; we don't want to spam
-		// connectable nodes
+	}
+}
+
+// permanentNodeManager tries to keep the Gateway's node list healthy. As long
+// as the Gateway has fewer than minNodeListSize nodes, it asks a random peer
+// for more nodes. It also continually pings nodes in order to establish their
+// connectivity. Unresponsive nodes are aggressively removed.
+func (g *Gateway) permanentNodeManager(closeChan chan struct{}) {
+	defer close(closeChan)
+
+	for {
+		// Wait 5 seconds so that a controlled number of peer requests are made
+		// to nodes in the node list.
 		select {
-		case <-time.After(10 * time.Minute):
+		case <-time.After(5 * time.Second):
 		case <-g.threads.StopChan():
+			// Gateway is shutting down, close the thread.
 			return
 		}
+
+		g.mu.RLock()
+		numNodes := len(g.nodes)
+		peer, err := g.randomPeer()
+		g.mu.RUnlock()
+		if err != nil {
+			// Most commonly the error indicates that there are no peers yet.
+			// Waiting through the next iteration gives the gateway time to
+			// bootstrap.
+			continue
+		}
+
+		// Determine whether there are a satisfactory number of nodes in the
+		// nodelist. If there are not, use the random peer from earlier to
+		// expand the node list.
+		if numNodes < minNodeListLen {
+			// TODO: should this be done in a go func? It's a blocking
+			// function.
```

Comment:
If it's not in a goroutine and we only have 1 peer then we would repeatedly call the RPC on the peer every 5 seconds. Probably best to keep it as a blocking call, but set a relatively short deadline in `g.requestNodes`.